### PR TITLE
[Service create form 2 of 2] First simple form with JSON editor

### PR DESCRIPTION
**Pull request by mlunoe**
This PR introduces the first working form, with interaction between JSON editor and the form section(s).

Here's a video showing it working:

Here's how to activate the new create service form:

### CODE REVIEW

**Comment by mlunoe** [comment]
Retest this please

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 1]
```diff
@@ -2,81 +2,200 @@ import Ace from 'react-ace';
...
+        return item.validationReducers;
+      }))
+    );
+
     this.state = {
-      jsonValue: ''
+      appConfig: {},
+      batch,
+      errors: {},
+      inputConfigReducers,
```
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Do we update reducers over the time? Otherwise `state` doesn't look like the right place to me.

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 2]
```diff
@@ -2,81 +2,200 @@ import Ace from 'react-ace';
...
 class NewCreateServiceModalForm extends React.Component {
   constructor() {
     super(...arguments);
 
+    let batch = new Batch();
+
+    let jsonParserReducers = ParserUtil.combineParsers(JSONParserReducers);
+    let jsonConfigReducers = ReducerUtil.combineReducers(JSONConfigReducers);
+    let inputConfigReducers = ReducerUtil.combineReducers(
+      Object.assign({}, ...SECTIONS.map((item) => {
```
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: less then 80 characters :)
```
Object.assign({}, ...SECTIONS.map((item) => item.configReducers))
```

File: plugins/services/src/js/reducers/JSONConfigReducers.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import ReducerUtil from '../../../../../src/js/utils/ReducerUtil';
```
Comment (nLight) [inline plugins/services/src/js/reducers/JSONConfigReducers.js]: We could deconstruct ReducerUtil a little
```
import {simpleReducer, simpleFloatReducer, simpleIntReducer} ...
```

File: src/js/utils/ReducerUtil.js [hunk 4]
```diff
@@ -53,7 +56,47 @@ module.exports = {
...
         // increase in performance.
-        state[key] = reducer.call(context.get(reducer), state[key], action);
+        localState[key] = reducer.call(context.get(reducer), localState[key], action);
+      }
+
+      return localState;
+    };
+  },
+
+  simpleReducer(needle, defaultState = '') {
```
Comment (nLight) [inline src/js/utils/ReducerUtil.js]: Would be awesome to have tests for this module :)

File: plugins/services/src/js/reducers/ServiceConfigReducers.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import ReducerUtil from '../../../../../src/js/utils/ReducerUtil';
```
Comment (nLight) [inline plugins/services/src/js/reducers/ServiceConfigReducers.js]: If you consider deconstructing, there's one more

File: plugins/services/src/js/reducers/JSONParserReducers.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import ParserUtil from '../../../../../src/js/utils/ParserUtil';
```
Comment (nLight) [inline plugins/services/src/js/reducers/JSONParserReducers.js]: If you consider deconstructing, there's one more

**Comment by Poltergeist** [comment]
There is one commit 07549460667cb11ce85d3d04a6510868a9ab995e which is left from the prototype which I would like to rename and also break apart and add tests for the simple reducer. Are you okay with me doing this, on this PR or do you want to do it @mlunoe?

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 7]
```diff
@@ -2,81 +2,200 @@ import Ace from 'react-ace';
...
+      this.setState({errors});
+    }
+  }
+
   handleJSONChange(jsonValue) {
-    this.setState({jsonValue});
+    try {
+      let batch = new Batch();
+      let parsedData = JSON.parse(jsonValue);
+      let appConfig = batch.reduce(this.state.jsonConfigReducers, parsedData);
```
Comment (Poltergeist) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Maybe we want to consider to do this outside of the try block because the try block will not be optimised.

File: src/js/structs/Batch.js [hunk 8]
```diff
@@ -42,7 +42,7 @@
 
 class Batch {
   constructor() {
-    const batch = [];
+    const batch = [{action: 'INIT'}];
```
Comment (Poltergeist) [inline src/js/structs/Batch.js]: This is a solution which I also considered to have the batch.reduce execute at least once. But I do think this is more a workaround then a solution to the problem. We should consider to rewrite the reduce function like that:

```JS
reduce(callback, data){
  if (this.length === 0) {
    return callback(data);
  }
  return this.reduce(callback, data);
}
```

In that case we are executing the callback at least once. And we are saving one cycle of `combineReducers` which does not any `value`.

Comment (Poltergeist) [inline src/js/structs/Batch.js]: Okay.

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 9]
```diff
@@ -2,81 +2,200 @@ import Ace from 'react-ace';
...
-              </TabView>
-            </TabViewList>
-          </Tabs>
+          <form onChange={this.handleFormChange} onBlur={this.handleFormBlur}>
+            <Tabs vertical={true}>
+              <TabButtonList>
+                <TabButton id="services" label="Services" />
+              </TabButtonList>
+              <TabViewList>
+                <ServiceFormSection errors={errors} data={data} />
```
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: I'll move the `TabView` to wrap the `ServiceFormSection`

Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Done

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 2 above)
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: done

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 1 above)
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: done

File: src/js/utils/ReducerUtil.js (hunk 4 above)
Comment (mlunoe) [inline src/js/utils/ReducerUtil.js]: Done!

File: plugins/services/src/js/reducers/ServiceConfigReducers.js (hunk 5 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/ServiceConfigReducers.js]: Done!

File: plugins/services/src/js/reducers/JSONParserReducers.js (hunk 6 above)
Comment (mlunoe) [inline plugins/services/src/js/reducers/JSONParserReducers.js]: Done!

**Comment by mlunoe** [comment]
All comments addressed @nLight and @Poltergeist!

**Comment by Poltergeist** [comment]
I rebased this since there where commits present which where already on Master.

**Comment by orlandohohmeier** [comment]
@mlunoe Can we merge this or do we need to land #1316 first?

**Comment by mlunoe** [comment]
@orlandohohmeier #1316 needs to land first :(

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
+import React, {Component} from 'react';
+import {Tooltip} from 'reactjs-components';
+
+import FieldInput from '../../../../../../src/js/components/form/FieldInput';
+import FieldLabel from '../../../../../../src/js/components/form/FieldLabel';
+import FieldHelp from '../../../../../../src/js/components/form/FieldHelp';
+import FieldError from '../../../../../../src/js/components/form/FieldError';
+import FormGroup from '../../../../../../src/js/components/form/FormGroup';
+import FieldTextarea from '../../../../../../src/js/components/form/FieldTextarea';
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: :abcd:

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+import Icon from '../../../../../../src/js/components/Icon';
+import ServiceConfigReducers from '../../reducers/ServiceConfigReducers';
+import ServiceValidationReducers from '../../reducers/ServiceValidationReducers';
+
+class ServiceFormSection extends Component {
+  getCMDLabel() {
+    let content = (
+      <span>
+        {"The command value will be wrapped by the underlying Mesos executor via /bin/sh -c ${cmd}. "}
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: #nit Single quotes... is our linter working?

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+          </h2>
+          <p>
+            Configure your service below. Start by giving your service a name.
+          </p>
+        </div>
+
+        <div className="flex row">
+          <FormGroup className="column-8" hasError={Boolean(errors.id)}>
+            <FieldLabel>
+              SERVICE NAME <span className="text-danger">*</span>
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: #nit Let's use title casing for these labels; the CSS should enforce `text-transform: uppercase`. That way, if we choose not to display uppercase labels, we can just change it in the CSS.

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+              name="id"
+              type="text"
+              value={data.id} />
+            <FieldHelp>{this.getIDHelpBlock()}</FieldHelp>
+            <FieldError>{errors.id}</FieldError>
+          </FormGroup>
+
+          <FormGroup className="column-4" hasError={Boolean(errors.instances)}>
+            <FieldLabel>
+              INSTANCES <span className="text-danger">*</span>
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: Title case here too

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+            <FieldInput
+              name="cpus"
+              type="number"
+              step="0.01"
+              value={data.cpus} />
+            <FieldError>{errors.cpus}</FieldError>
+          </FormGroup>
+          <FormGroup className="column-4" hasError={Boolean(errors.mem)}>
+            <FieldLabel>
+              MEMORY (MiB) <span className="text-danger">*</span>
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: What do you think about handling the asterisk inside the `FieldLabel` component? We could pass an `isRequired` prop to `FieldLabel` and then it knows to append this asterisk...

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+              type="number"
+              value={data.disk} />
+            <FieldError>{errors.disk}</FieldError>
+          </FormGroup>
+        </div>
+
+        <FormGroup hasError={Boolean(errors.cmd)}>
+          <FieldTextarea
+            error={errors.cmd}
+            helpBlock="A shell command for your container to execute."
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: I think this `helpBlock` prop should be a separate component?

File: plugins/services/src/js/components/forms/ServiceFormSection.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+              MEMORY (MiB) <span className="text-danger">*</span>
+            </FieldLabel>
+            <FieldInput
+              name="mem"
+              type="number"
+              value={data.mem} />
+            <FieldError>{errors.mem}</FieldError>
+          </FormGroup>
+          <FormGroup className="column-4" hasError={Boolean(errors.disk)}>
+            <FieldLabel>DISK (MiB)</FieldLabel>
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: Title case here :)

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 14 above)
Comment (jfurrow) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: Also title case this label plz :)

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 17]
```diff
@@ -2,81 +2,207 @@ import Ace from 'react-ace';
 import classNames from 'classnames';
 import React from 'react';
 
-import Tabs from '../../../../../../src/js/components/Tabs';
+import Batch from '../../../../../../src/js/structs/Batch';
+import JSONConfigReducers from '../../reducers/JSONConfigReducers';
+import JSONParserReducers from '../../reducers/JSONParserReducers';
+import {combineParsers} from '../../../../../../src/js/utils/ParserUtil';
+import {combineReducers} from '../../../../../../src/js/utils/ReducerUtil';
```
Comment (jfurrow) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: :abcd:

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 18]
```diff
@@ -2,81 +2,207 @@ import Ace from 'react-ace';
...
+      newState.errors = Object.assign(errors, {
+        jsonEditor: 'JSON value is not valid json.'
+      });
+    }
+
+    if (appConfig) {
+      // Flush batch
+      let batch = new Batch();
+
+      // Run validation reducers on appConfig with new on minimal batch
```
Comment (jfurrow) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: Could you rephrase this comment? I'm struggling to figure out what you mean by`with new on minimal batch`

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 19]
```diff
@@ -2,81 +2,207 @@ import Ace from 'react-ace';
...
+      let batch = new Batch();
+      let appConfig = batch.reduce(jsonConfigReducers, parsedData);
+      Object.assign(newState, {batch, appConfig});
+    }
+
+    this.setState(newState);
+  }
+
+  handleFormBlur() {
+
```
Comment (jfurrow) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: #nit let's kill this newline

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 11 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: hmmm... weird... maybe not?

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 10 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: done

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 11 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: done

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 13 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: done

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 16 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: done

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 17 above)
Comment (mlunoe) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: done

File: plugins/services/src/js/components/forms/ServiceFormSection.js (hunk 14 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/ServiceFormSection.js]: Ok, implemented abilities to add required indicator and decide whether to display the error field

**Comment by mlunoe** [comment]
Merging this guy after approval from @jfurrow @orlandohohmeier and @Poltergeist
